Add Package Manager Friendly Start Up Script

schrieveslaach opened the pull request:
Introducing a package manager friendly start up script to make the handling of jdt.ls easier (fixes #1823 and fixes #1934).

Why did I choose Python and not bash? Basically to address the [following concern](https://github.com/eclipse/eclipse.jdt.ls/issues/1823#issuecomment-899779358):

>  The tricky part is getting and testing something for Windows, Mac, Linux.

The script has been written to be platform agnostic so that it can be used on Window, Linux, and Mac. So there is no need to write separate scripts for each platform.

It also adds some logic to bootstrap Lombok out of the box. This feature could be hidden behind a command line option.

schrieveslaach commented:
@languitar, as you are the maintainer of [the AUR jdtls version](https://aur.archlinux.org/packages/jdtls) and I'm trying to bring it to [Homebrew](https://github.com/Homebrew/homebrew-core/pull/92550), your feedback is highly appreciated (at least from my point of view).

languitar commented:
On first sight this looks good. Apart from the magic of finding java etc. this looks a lot like what I use in my AUR package.

mfussenegger reviewed (commented):
I'm not a maintainer but noticed a couple of things:

- I think `python` might still refer to python2 on some systems, which would break because  at least `pathlib` wouldn't be available
- The `Running from the command line` section in the readme would probably need an update.
- I think on windows it is still not uncommon to not have python at all?
- This would remove the option to easily configure heap size and GC. Not sure if that's a common use case, just want to point it out.
- I suspect the version check would fail if using early access releases. Not sure if this is common enough to support - but just to mention it.


```
❮ ./.local/jdks/jdk-17ea-panama/bin/java -version
openjdk version "17-panama" 2021-09-14
OpenJDK Runtime Environment (build 17-panama+3-167)
OpenJDK 64-Bit Server VM (build 17-panama+3-167, mixed mode, sharing)

~
❮ ./.local/jdks/jdk-19ea-panama/bin/java -version
openjdk version "19-panama" 2022-09-20
OpenJDK Runtime Environment (build 19-panama+1-13)
OpenJDK 64-Bit Server VM (build 19-panama+1-13, mixed mode, sharing)
```
I am not sure about the output of builds other than OpenJDK? 


With that in mind - I think a start-up script would in general be a nice addition.

schrieveslaach commented:
@mfussenegger, @languitar, thanks for your feedback. I'll look into this.

I already added `#!/usr/bin/env python3` to ensure that Python 3 will be used.

languitar commented on org.eclipse.jdt.ls.product/scripts/jdtls:
```diff
@@ -0,0 +1,110 @@
+	settings_xml = os.path.join(Path.home(), '.m2', 'settings.xml')
+
+	if os.path.exists(settings_xml):
+		settings = ET.parse(settings_xml)
```
Is it safe to use ElementTree here to parse this file with the potential for injecting malicious XML? From the docs:

> The [xml.etree.ElementTree](https://docs.python.org/3/library/xml.etree.elementtree.html#module-xml.etree.ElementTree) module is not secure against maliciously constructed data. If you need to parse untrusted or unauthenticated data see [XML vulnerabilities](https://docs.python.org/3/library/xml.html#xml-vulnerabilities).

https://docs.python.org/3/library/xml.etree.elementtree.html

schrieveslaach replied:
Switched to `defusedxml`.

languitar commented on org.eclipse.jdt.ls.product/scripts/jdtls:
```diff
@@ -0,0 +1,110 @@
+	elif system == 'Darwin':
+		config_dir = 'config_mac'
+	elif system == 'Windows':
+		config_dir = 'config_win'
```
```suggestion
		config_dir = 'config_win'
	else:
		raise Exception("Unknown platform {} detected".format(platform))
```

Otherwise this would fail with a cryptic error that config_dir is not defined.

schrieveslaach replied:
Done.

languitar commented on org.eclipse.jdt.ls.product/scripts/jdtls:
```diff
@@ -0,0 +1,110 @@
+		if os.path.exists(java_exec_to_test):
+			java_executable = java_exec_to_test
+
+	p = subprocess.Popen(java_executable + ' -version', shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE, universal_newlines=True)
```
Why not [`check_output`](https://docs.python.org/3/library/subprocess.html#subprocess.check_output)? Should be easier to use.

schrieveslaach replied:
Yes, much better option. Thanks for the hint.

schrieveslaach commented:
@languitar, @mfussenegger, I've updated this PR and I took your comments into account. Do you mind to have another look?

languitar commented on org.eclipse.jdt.ls.product/scripts/jdtls:
```diff
@@ -0,0 +1,118 @@
+			java_executable = java_exec_to_test
+
+	if validate_java_version:
+		out = subprocess.check_output(java_executable + ' -version', shell = True, stderr = subprocess.STDOUT, universal_newlines=True)
```
I think I asked that before: why with `shell=True`? That only gives the potential for injections.

```suggestion
		out = subprocess.check_output([java_executable, '-version'], stderr = subprocess.STDOUT, universal_newlines=True)
```

schrieveslaach replied:
Sorry, I did not notice the suggestion last time. I'll cover it.

languitar reviewed (commented):
I don't know which minimum version of Python you want to target. But for modern (tm) Python one would probably migrate all the os.path calls to pathlib.

rgrunber reviewed (commented):
Many users have noted this would be really good to have, so I think it's worth adding. @testforstephen let me know what you think about this.

Note that the script is not currently packaged in the final tarball at `org.eclipse.jdt.ls.product/distro/jdt-language-server-1.9.0-202202221741.tar.gz`, but I guess this is fine as it's meant as more a developer utility.

testforstephen reviewed (changes requested):
I understand adding a startup script would bring some value. As a Java developer, I'm more used to add wrapper scripts using maven wrapper way, that's more clean and no extra dependencies. 

However, if the target audience of this PR is mainly for the automated Package Manager, using python might be fine. To keep the implementation simple, i'd like to remove the extra python dependencies for `defusedxml` and `semver`. And also don't handle lombok.jar in the startup of language server, since Lombok support is out of scope of eclipse.jdt.ls project.

schrieveslaach commented on org.eclipse.jdt.ls.product/scripts/jdtls:
```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env python3
```
I think [that approach](https://stackoverflow.com/a/3592208/5088458) should be used so that it is possible to use it on Windows and Linux by just typing `jdtls` in the command line.

schrieveslaach commented:
I updated the PR and addressed the suggestions

> But for modern (tm) Python one would probably migrate all the os.path calls to pathlib.

@languitar, moved to pathlib

> Note that the script is not currently packaged in the final tarball at org.eclipse.jdt.ls.product/distro/jdt-language-server-1.9.0-202202221741.tar.gz, but I guess this is fine as it's meant as more a developer utility.

@rgrunber, actually I intended to ensure that the script will be part of the final `tar.gz` because it is requested by the homebrew maintainers (see [here](https://github.com/Homebrew/homebrew-core/pull/92550#issuecomment-1019868246). I updated the PR so that it is included now.

> To keep the implementation simple, i'd like to remove the extra python dependencies for defusedxml and semver. And also don't handle lombok.jar in the startup of language server, since Lombok support is out of scope of eclipse.jdt.ls project.

@testforstephen, I remove the Python packages and removed the Lombok option (I think I'll write a Neovim plugin to resolve Lombok that can be used in conjunction with @mfussenegger's `nvim-jdtls`).

I've added some other command line option to specify additional Java properties to be able to specify the Lombok Java agent.

> … since Lombok support is out of scope of eclipse.jdt.ls project.

@testforstephen, is it completely out of scope (I'm asking out of curiosity)? I've discovered an issues while loading the lombok Java agent and I would like to report those.

schrieveslaach commented on README.md:
```diff
@@ -67,21 +67,29 @@ java \
-
 If you want to debug eclipse.jdt.ls itself, add `-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=1044` right after `java` and ensure nothing else is running on port 1044. If you want to debug from the start of execution, change `suspend=n` to `suspend=y` so the JVM will wait for your debugger prior to starting the server.
 
+There is also a python wrapper script availabe that makes the start up of eclipse.jdt.ls more convenient (no need to juggle with Java options etc.). You can use following command to start eclipse.jdt.ls through the script if on your system is [`defusedxml`](https://pypi.org/project/defusedxml/) and [`semver`](https://pypi.org/project/semver/) installed:
```
Those dependencies have been removed.

schrieveslaach commented on org.eclipse.jdt.ls.product/publish-assembly.xml:
```diff
@@ -9,8 +9,9 @@
     <fileSet>
       <directory>${basedir}/target/repository</directory>
-      <outputDirectory>/</outputDirectory>
+	  <outputDirectory/>
```
Removes a warning: https://stackoverflow.com/a/28506199/5088458

rgrunber commented on org.eclipse.jdt.ls.product/scripts/jdtls.py:
```diff
@@ -0,0 +1,93 @@
+import argparse
```
As this is a source file (particularly one we'll be shipping from Eclipse servers (eg. https://download.eclipse.org/jdtls/milestones/1.8.0/, https://repo.eclipse.org/content/repositories/jdtls-releases/org/eclipse/jdt/ls/org.eclipse.jdt.ls.product/1.8.0.20220126150144/org.eclipse.jdt.ls.product-1.8.0.20220126150144.tar.gz), we should add a copyright header here much like we do for Java sources (obviously the comment declaration will be a bit different, with `"""`).

See https://www.eclipse.org/projects/handbook/#ip-copyright-headers for the specific requirement.

Otherwise, change looks good to me.

schrieveslaach replied:
I've add the headers, hopefully in the correct way. Can you double check?

rgrunber replied:
Yes, just update the years to 2022 and `SPDX-License-Identifier` should be `EPL-2.0` (no 3 at end) as per https://spdx.org/licenses/EPL-2.0.html .

schrieveslaach replied:
Sorry, stupid copy paste error on my side. 😅 Fixed.

languitar commented on org.eclipse.jdt.ls.product/scripts/jdtls.py:
```diff
@@ -0,0 +1,93 @@
+			jvm_options = " ".join(f"'{w}'" for w in known_args.jvm_arg),
+			args = " ".join(f"'{w}'" for w in args)))
+
+sys.modules[__name__] = jdtls
```
What's the idea here? I've never seen that. And why not something like

```python
if __name__ = "__main__":
    jdtls()
```

schrieveslaach replied:
[As stated above](https://github.com/eclipse/eclipse.jdt.ls/pull/2005#discussion_r813820478) I followed [an approach](https://stackoverflow.com/questions/2347042/how-to-run-a-python-script-without-specifying-the-file-extension-cross-platform/3592208#3592208) so that it is possilbe to run the script by typing just `jdtls` on Linux, Mac, and Windows. However, I removed that line and improved the code a bit.

languitar commented on org.eclipse.jdt.ls.product/scripts/jdtls.py:
```diff
@@ -0,0 +1,93 @@
+	out = subprocess.check_output([java_executable, '-version'], stderr = subprocess.STDOUT, universal_newlines=True)
+
+	matches = re.finditer(r"(?P<major>\d+)\.\d+\.\d+", out)
+	for matchNum, match in enumerate(matches):
+		java_major_version = int(match.group("major"))
```
```suggestion
	for match in matches:
		java_major_version = int(match.group("major"))
```

`matchNum` is never used

languitar commented on org.eclipse.jdt.ls.product/scripts/jdtls.py:
```diff
@@ -0,0 +1,93 @@
+
+def find_equinox_launcher(jdtls_base_directory):
+	plugins_dir = jdtls_base_directory / "plugins"
+	launchers = glob.glob('org.eclipse.equinox.launcher_*.jar', root_dir = plugins_dir)
```
```suggestion
	launchers = plugins_dir.glob('org.eclipse.equinox.launcher_*.jar')
```

This would also eliminate the glob import

schrieveslaach replied:
Needed to change the `if` to a `for` loop because `Path.glob` returns a generator object.

languitar reviewed (commented):
Some minor things I still see.

rgrunber commented:
@testforstephen looks like issues have been addressed here. Have a look and let me know if you're fine with merging.

rgrunber commented:
@schrieveslaach thanks for taking the time to address all the suggestions.

schrieveslaach commented:
Thanks for merging. 🙂

milind-shakya-sp commented on org.eclipse.jdt.ls.product/scripts/jdtls.py:
```diff
@@ -0,0 +1,102 @@
+
+def main(args):
+	parser = argparse.ArgumentParser()
+	parser.add_argument("--validate-java-version", default=True, action=argparse.BooleanOptionalAction)
```
`argparse.BooleanOptionalAction` is only available in python 3.9. So essentially everyone using python version < 3.9, get the above error.

schrieveslaach replied:
Issue has been raised in #2153
